text/0000-router-service.md

```diff
@@ -0,0 +1,440 @@
+
+1. Here is a simplified `is-active` helper that will always update at the appropriate time to match exactly what is rendered in the current outlet. It will maintain the correct state even during animations. Instead of injecting the router service, it consumes the `routeInfo` from its containing environment:
+
+```js
+Ember.Helper.extend({
+  compute([routeName], { routeInfo }) {
+    return !!routeInfo.find(info => info.name === routeName);
+  }
+}).reopenClass({
+  getDynamicVariables: ['routeInfo']
+});
+```
+
+A more complete version that also matches models and queryParams can be written in the same way. 
+
+2. We can improve `link-to` so that it always finds implicit model arguments from the local context, rather than trying to locate them on the global router service. This will fix longstanding bugs like https://github.com/ember-animation/liquid-fire/issues/347 and it will make it easier to test components that contain `{{link-to}}`. This would also open the door to relative link-tos.
+
+3. `liquid-outlet` can be implemented entirely via public API. It would become:
+
+```hbs
+{{#liquid-bind (get-dynamic-variable "routeInfo") as |currentRouteInfo|}}
+  {{#with-dynamic-variable "routeInfo" currentRouteInfo}}
+    {{outlet}}
+  {{/with-dynamic-variable}}
+{{/liquid-bind}}
+```
+
+4. Prerendering of non-current routes becomes possible. You can use `recognizeAndLoad` to obtain a `RouteInfoWithAttributes` and then use `{{#with-dynamic-variable "routeInfo" myRouteInfo}} {{outlet}} {{/with-dynamic-variable}}` to render it.
+
+
+# Drawbacks
+
+This RFC deprecates only two public extension hooks API, so the API-churn burden may appear low. However, we know that use of the private APIs we're deliberately disabling is widespread, so users will experience churn. We can provide our usual deprecation cycle to give them early warning, but it still imposes some cost.
+
+This RFC doesn't attempt to change the existing and fairly rich semantics for initiating transitions. For example, you can pass either models or IDs, and those have subtle semantic differences. I think an ideal rewrite would also change the semantics of the route hooks and transitionTo to simplify that area.
+
+# Alternatives
+
+## Less Churn
+
+We could adopt some of the existing broadly used APIs as de-facto public. This avoids churn, but imposes a complexity burden on every new learner, who needs to be told "this is a weird API, but it's what we're stuck with".
+
+## Semver Lawyering
+
+I'm interpreting router.js's public/private documentation as out-of-scope for Ember's semver. The fact that we pass an instance of router.js's Transition as our `transition` argument is not documented. An alternative interpretation is that we need to continue supporting those methods marked as public in router.js's docs.
```

Comment:
`router.js` is on my list to tackle; in my approach to `route-recognizer` I treated anything that I could get away with without breaking Ember's public and intimate APIs (turned out to be very little, but whatever) as fair game to change. I'd like to take the same approach here. Breaking change to `router.js`; SemVer compatible change to Ember.